build/source/driver/multi_driver.f90

```diff
@@ -239,79 +263,102 @@ real(dp),parameter        :: safety = 0.01_dp               ! safety factor to e
-! (3) read information for each HRU and allocate space for data structures
+! (3a) read the number of GRUs and HRUs, and allocate the gru-hru mapping structures
+! *****************************************************************************
+
+! read and allocate for gru-hru mapping structures and get global variables, e.g, nGRU, nHRU,
+! needed in the consequent allocations
+! nGRU-is the total number of GRUs of the simulation domain
+! nHRU-is the total number of HRUs of the simulation domain
+! hruCount-is a local variable for the total number of HRUs in a GRU
+call allocate_gru_struc(nGRU,nHRU,err,message); call handle_err(err,message)
+
+! *****************************************************************************
+! (3b) read the number of snow and soil layers
 ! *****************************************************************************
 
 ! *** TEMPORARY CODE ***
 ! code will be replaced once merge with the NetCDF branch
 
-! get the number of HRUs
-call file_open(trim(SETNGS_PATH)//trim(LOCAL_ATTRIBUTES),fileUnit,err,message); call handle_err(err,message)
-call get_vlines(fileUnit,dataLines,err,message); call handle_err(err,message)
-nHRU = size(dataLines)-1  ! -1 because of the header
-deallocate(dataLines)
-close(fileUnit)
-
-! allocate space for the number of snow and soil layers
-allocate(nSnow(nHRU),nSoil(nHRU),stat=err); call handle_err(err,'unable to allocate space for the number of snow and soil layers')
-
-! loop through HRUs
-! NOTE: currently the same initial conditions for all HRUs; will change when shift to NetCDF
-do iHRU=1,nHRU
-
- ! get a vector of non-commented lines
- call file_open(trim(SETNGS_PATH)//trim(MODEL_INITCOND),fileUnit,err,message); call handle_err(err,message)
- call get_vlines(fileUnit,dataLines,err,message); call handle_err(err,message)
- close(fileUnit)
-
- ! get the number of snow and soil layers for each HRU
- nSnow(iHRU)=0           ! initialize the number of snow layers
- nSoil(iHRU)=0           ! initialize the number of soil layers
- do iVar=1,size(dataLines)
-  ! split the line into an array of words
-  call split_line(dataLines(iVar),chardata,err,message); call handle_err(err,message)
-  ! check if the line contains initial conditions data (contains the word "snow" or "soil")
-  do iword=1,size(chardata)
-   if(chardata(iword)=='snow') nSnow(iHRU) = nSnow(iHRU)+1
-   if(chardata(iword)=='soil') nSoil(iHRU) = nSoil(iHRU)+1
-   if(chardata(iword)=='snow' .or. chardata(iword)=='soil') exit ! exit once read the layer type
+! NOTE: currently the same initial conditions for all HRUs and GRUs; will change when shift to NetCDF
+
+! check the GRU-HRU mapping structure is allocated
+if(.not.allocated(gru_struc)) call handle_err(err,'gru_struc is not allocated')
+
+! loop through GRUs
+do iGRU=1,nGRU
+
+ ! check the GRU-HRU mapping structure is allocated 
+ if(.not.allocated(gru_struc(iGRU)%hruInfo)) call handle_err(err,'gru_struc(iGRU)%hruInfo is not allocated')
```

Comment:
Done (for the high-level allocation check).

Note that the code in question here is temporary code, which will be removed when we upgrade to netcdf initial conditions (issue #125)